lib/elixir/lib/module/types/descr.ex

```diff
@@ -495,4 +525,562 @@ defmodule Module.Types.Descr do
+      map_get_static(descr, key)
+    else
+      {dynamic, static} = Map.pop(descr, :dynamic)
+      dynamic_value_type = map_get_static(dynamic, key)
+      static_value_type = map_get_static(static, key)
+      union(intersection(dynamic(), dynamic_value_type), static_value_type)
+    end
+  end
+
+  @doc """
+  Check that a key is present.
+  """
+  def map_has_key?(descr, key) do
+    subtype?(descr, map([{key, term()}], :open))
+  end
+
+  # Assuming `descr` is a static type. Accessing `key` will, if it succeeds,
+  # return a value of the type returned. To guarantee that the key is always
+  # present, use `map_has_key?`. To guarantee that the key may be present
+  # use `map_may_have_key?`. If key is never present, result will be `none()`.
+  defp map_get_static(descr, key) when is_atom(key) do
+    descr_map = intersection(descr, map())
+
+    if empty?(descr_map) do
+      none()
+    else
+      map_split_on_key(descr_map.map, key)
+      |> Enum.reduce(none(), fn {typeof_key, _}, union -> union(typeof_key, union) end)
+      |> remove_not_set()
+    end
+  end
+
+  @doc """
+  Check that a key may be present.
+  """
+  def map_may_have_key?(descr, key) do
+    compatible?(descr, map([{key, term()}], :open))
+  end
+
+  # Union is list concatenation
+  defp map_union(dnf1, dnf2), do: dnf1 ++ dnf2
+
+  defp map_intersection(dnf1, dnf2) do
+    case Enum.flat_map(dnf1, &map_intersection_aux(&1, dnf2, [])) do
+      [] -> 0
+      x -> x
+    end
+  end
+
+  defp map_intersection_aux(_, [], acc), do: acc
+
+  defp map_intersection_aux({pos1, negs1}, [{pos2, negs2} | rest_dnf2], acc) do
+    try do
+      x = map_literal_intersection(pos1, pos2)
+      map_intersection_aux({pos1, negs1}, rest_dnf2, [{x, negs1 ++ negs2} | acc])
+    catch
+      :empty_intersection -> map_intersection_aux({pos1, negs1}, rest_dnf2, acc)
+    end
+  end
+
+  # Creates the intersection, if not empty, of two map literals.
+  defp map_literal_intersection(map1, map2) do
+    case {map1, map2} do
+      {{:closed, fs1}, {:closed, fs2}} -> {:closed, merge_closed(fs1, fs2)}
+      {{:open, fs1}, {:closed, fs2}} -> {:closed, merge_open_with_closed(fs1, fs2)}
+      {{:closed, fs1}, {:open, fs2}} -> {:closed, merge_open_with_closed(fs2, fs1)}
+      {{:open, fs1}, {:open, fs2}} -> {:open, merge_open(fs1, fs2)}
+    end
+  end
+
+  defp go_through_keys([], []), do: []
+  defp go_through_keys([], rest), do: Enum.map(rest, &{:right, &1})
+  defp go_through_keys(rest, []), do: Enum.map(rest, &{:left, &1})
+
+  defp go_through_keys([key1 | rest1], [key2 | rest2]) do
+    cond do
+      key1 < key2 -> [{:left, key1} | go_through_keys(rest1, [key2 | rest2])]
+      key1 > key2 -> [{:right, key2} | go_through_keys([key1 | rest1], rest2)]
+      true -> [{:both, key1} | go_through_keys(rest1, rest2)]
+    end
+  end
+
+  defp merge_closed(fields1, fields2) do
+    # if they don't have the same keys, intersection is empty
+    if Map.keys(fields1) != Map.keys(fields2) do
```

Comment:
Is it safe to assume the keys have the same order?

Reply:
Good point. Even if it happens to be the case now - not 100% sure - history has told us not to assume anything about key order 😅

Reply:
It is not guaranteed, not for large maps anyway. The best is to fold over one pair of keys, we can work on these optimizations soon.